src/externalDevice/BleChessProtocol.ts

```diff
@@ -1,22 +1,116 @@
 import { BaseProtocol, BaseState } from './BaseProtocol'
-import { isUciWithPromotion, isCentralStateCreated, createFullFen, lastMoveToUci, getCommandParams, sendCommandToPeripheral, sendMoveToCentral, areFensSame, sendStateChangeToCentral, applyPeripheralMoveRejected, applyPeripheralLastMove, applyPeripheralSynchronized, applyPeripheralPieces } from './utils'
+import { isCentralStateCreated, createFullFen, lastMoveToUci, getCommandParams, sendCommandToPeripheral, sendMoveToCentral, sendStateChangeToCentral, applyPeripheralMoveRejected, applyPeripheralLastMove, applyVariantSupported, applyPeripheralSynchronized, applyPeripheralPieces, createIterator } from './utils'
 import { State, makeDefaults } from '../chessground/state'
+import { GameStatus } from '../lichess/interfaces/game'
 import { Toast } from '@capacitor/toast'
 import i18n from '../i18n'
 
 export class BleChessProtocol extends BaseProtocol {
   roundState = makeDefaults()
-  features = new BleChessFeatures
+  features = new Features
+  variants = new Variants
+  variantsMap = {
+    standard: this.variants.standard,
+    chess960: this.variants.chess960,
+    antichess: this.variants.antiChess,
+    kingOfTheHill: this.variants.kingOfTheHill,
+    threeCheck: this.variants.threeCheck,
+    atomic: this.variants.atomic,
+    horde: this.variants.horde,
+    racingKings: this.variants.racingKings,
+    crazyhouse: this.variants.crazyHouse,
+  }
+  endReasonsMap = {
+    mate: EndReason.Checkmate,
+    stalemate: EndReason.Stalemate,
+    draw: EndReason.Draw,
+    timeout: EndReason.Timeout,
+    outoftime: EndReason.Timeout,
+    resign: EndReason.Resign,
+    aborted: EndReason.Abort,
+    noStart: EndReason.Undefined,
+    unknownFinish: EndReason.Undefined,
+    cheat: EndReason.Undefined,
+    variantEnd: EndReason.Undefined,
+  }
 
   init(st: State) {
     this.roundState = st
     this.transitionTo(new Init)
   }
 }
 
-class BleChessFeatures {
-  msg: boolean = false
-  lastMove: boolean = false
+enum FeatureName {
+  Msg = 'msg',
+  LastMove = 'last_move',
+}
+
+enum VariantName {
+  Standard = "standard",
+  Chess960 = "chess_960",
+  ThreeCheck = "3_check",
+  Atomic = "atomic",
+  KingOfTheHill = "king_of_the_hill",
+  AntiChess = "anti_chess",
+  Horde = "horde",
+  RacingKings = "racing_kings",
+  CrazyHouse = "crazy_house",
+}
+
+enum Command {
+  Ok = 'ok',
+  Nok = 'nok',
+  Feature = 'feature',
+  Variant = 'variant',
+  SetVariant = 'set_variant',
+  Begin = 'begin',
+  State = 'state',
+  Sync = 'sync',
+  Unsync = 'unsync',
+  End = 'end',
+  Move = 'move',
+  Promote = 'promote',
+  Check = 'check',
+  Err = 'err',
+  Msg = 'msg',
+  LastMove = 'last_move',
+}
+
+enum EndReason {
+  Undefined = 'undefined',
```

Comment:
what about:
 - cheatdetected

Reply:
`undefined` is more common, not sure if we need additional `cheat` reason